### ISSUE TITLE
请问是不是models.resnet这些网络的实现有问题？

### ISSUE THREAD

**Issue by AmigoCDT** [weiaicunzai/pytorch-cifar100#17]
我的pytorch==1.2, torchvision=0.4.0
我试着用torchvision.models中实现的ResNet去运行train.py训练，结果在测试集上的ACC只有 ***60%*** 左右。我用作者你实现的ResNet训练的话，ACC大概符合预期。
二者区别还有一个，Batch_size=128时，用torchvision.models中的网络训练，GPU_Memory大概只用了1.4G，而用作者你的实现，GPU_Memory大概要占用5G左右，感觉像是给Batch-Size中每个图片都分配了一个model。
请问这个实现网络的时候有区别，还是咱们的pytorch版本不同导致的问题？
谢谢。

**Comment by qiulinzhang** [comment]
我也有这个问题，作者的实现的差不多5个G, pytorch官方实现差不多1.4个G，请问有看到问题在哪吗？

**Comment by zxj-123** [comment]
> 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> 而且pytorch官方实现速度快很多

你用的resnet多少啊

**Comment by zxj-123** [comment]
> 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> 而且pytorch官方实现速度快很多
torchvision没有cifar100的resnet啊

**Comment by qiulinzhang** [comment]
> > 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> > 而且pytorch官方实现速度快很多
> 
> 你用的resnet多少啊

resent50

**Comment by zxj-123** [comment]
> 我的pytorch==1.2, torchvision=0.4.0
> 我试着用torchvision.models中实现的ResNet去运行train.py训练，结果在测试集上的ACC只有 _**60%**_ 左右。我用作者你实现的ResNet训练的话，ACC大概符合预期。
> 二者区别还有一个，Batch_size=128时，用torchvision.models中的网络训练，GPU_Memory大概只用了1.4G，而用作者你的实现，GPU_Memory大概要占用5G左右，感觉像是给Batch-Size中每个图片都分配了一个model。
> 请问这个实现网络的时候有区别，还是咱们的pytorch版本不同导致的问题？
> 谢谢。



> > > 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> > > 而且pytorch官方实现速度快很多
> > 
> > 
> > 你用的resnet多少啊
> 
> resent50

我有一点疑惑，作者这个不是cifar100数据集的吗，torchvision.models不是imagent数据集的吗。

**Comment by qiulinzhang** [comment]
> > 我的pytorch==1.2, torchvision=0.4.0
> > 我试着用torchvision.models中实现的ResNet去运行train.py训练，结果在测试集上的ACC只有 _**60%**_ 左右。我用作者你实现的ResNet训练的话，ACC大概符合预期。
> > 二者区别还有一个，Batch_size=128时，用torchvision.models中的网络训练，GPU_Memory大概只用了1.4G，而用作者你的实现，GPU_Memory大概要占用5G左右，感觉像是给Batch-Size中每个图片都分配了一个model。
> > 请问这个实现网络的时候有区别，还是咱们的pytorch版本不同导致的问题？
> > 谢谢。
> 
> > > > 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> > > > 而且pytorch官方实现速度快很多
> > > 
> > > 
> > > 你用的resnet多少啊
> > 
> > 
> > resent50
> 
> 我有一点疑惑，作者这个不是cifar100数据集的吗，torchvision.models不是imagent数据集的吗。

数据集这个应该没什么关系吧，模型都是一样的，只不过不同的数据集数据处理方式归一化等不一样

**Comment by Lornatang** [comment]
The implementation of CIFAR in the original ResNet paper is reproduced in my repository, please see (https://github.com/Lornatang/ResNet/tree/master/examples/cifar)
Thank you, please give me some suggestions. @AmigoCDT

**Comment by KseeleK** [comment]
> 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> 而且pytorch官方实现速度快很多

我自己用官方的ResNet-18（改了下第一层卷积核大小适应cifar数据集尺寸）在cifar100的准确率是65%~70%之间。
我也尝试完全照搬作者的实验和整个工程去在cifar100上训练resnet18，但准确率只有66%，完全达不到>70%的准确率

不知道有没有和我一样情况的？无法复现作者的准确率？

我目前认为Cifar100只做数据增广，而不采用一些trick是无法让准确率上70%的

**Comment by zxj-123** [comment]
> > 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> > 而且pytorch官方实现速度快很多
> 
> 我自己用官方的ResNet-18（改了下第一层卷积核大小适应cifar数据集尺寸）在cifar100的准确率是65%~70%之间。
> 我也尝试完全照搬作者的实验和整个工程去在cifar100上训练resnet18，但准确率只有66%，完全达不到>70%的准确率
> 
> 不知道有没有和我一样情况的？无法复现作者的准确率？
> 
> 我目前认为Cifar100只做数据增广，而不采用一些trick是无法让准确率上70%的

我只用了作者的model实现代码，可以达到他的准确率的

**Comment by xyy1991** [comment]
我按照作者的方法，也只得到了不到70%的准确率，请问大家解决了这个问题吗？

**Comment by xyy1991** [comment]
> > 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> > 而且pytorch官方实现速度快很多
> 
> 我自己用官方的ResNet-18（改了下第一层卷积核大小适应cifar数据集尺寸）在cifar100的准确率是65%~70%之间。
> 我也尝试完全照搬作者的实验和整个工程去在cifar100上训练resnet18，但准确率只有66%，完全达不到>70%的准确率
> 
> 不知道有没有和我一样情况的？无法复现作者的准确率？
> 
> 我目前认为Cifar100只做数据增广，而不采用一些trick是无法让准确率上70%的

请问您解决这个问题了吗？我好想遇到了跟你一样的情况

**Comment by KseeleK** [comment]
> > > 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> > > 而且pytorch官方实现速度快很多
> > 
> > 
> > 我自己用官方的ResNet-18（改了下第一层卷积核大小适应cifar数据集尺寸）在cifar100的准确率是65%~70%之间。
> > 我也尝试完全照搬作者的实验和整个工程去在cifar100上训练resnet18，但准确率只有66%，完全达不到>70%的准确率
> > 不知道有没有和我一样情况的？无法复现作者的准确率？
> > 我目前认为Cifar100只做数据增广，而不采用一些trick是无法让准确率上70%的
> 
> 请问您解决这个问题了吗？我好想遇到了跟你一样的情况

我自己的问题解决了，我是因为pytorch版本原因导致了学习率没有正常更新，你检查下你的学习率是否在训练过程中更新，如果没有，尝试下把
scheduler.step(epoch) 
改成
scheduler.step()

**Comment by xyy1991** [comment]
多谢哈，我这边也解决了，但是我发现是我的数据增强没有用上才导致精确度上不去的




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"WEIHAN LI"<notifications@github.com&gt;;
发送时间:&nbsp;2020年6月10日(星期三) 下午5:54
收件人:&nbsp;"weiaicunzai/pytorch-cifar100"<pytorch-cifar100@noreply.github.com&gt;;
抄送:&nbsp;"夏洋洋"<1319230184@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [weiaicunzai/pytorch-cifar100] 请问是不是models.resnet这些网络的实现有些问题？ (#17)





    
我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
 而且pytorch官方实现速度快很多
  
我自己用官方的ResNet-18（改了下第一层卷积核大小适应cifar数据集尺寸）在cifar100的准确率是65%~70%之间。
 我也尝试完全照搬作者的实验和整个工程去在cifar100上训练resnet18，但准确率只有66%，完全达不到&gt;70%的准确率
 不知道有没有和我一样情况的？无法复现作者的准确率？
 我目前认为Cifar100只做数据增广，而不采用一些trick是无法让准确率上70%的
  
请问您解决这个问题了吗？我好想遇到了跟你一样的情况
  
我自己的问题解决了，我是因为pytorch版本原因导致了学习率没有正常更新，你检查下你的学习率是否在训练过程中更新，如果没有，尝试下把
 scheduler.step(epoch)
 改成
 scheduler.step()

**Comment by animesh-007** [comment]
> 多谢哈，我这边也解决了，但是我发现是我的数据增强没有用上才导致精确度上不去的

您能告诉我如何解决这个问题，因为我无法获得相同的数字吗？

**Comment by animesh-007** [comment]
> 我自己实现的时候没有用dataloader，datasets虽然有transform这个属性，但是并没有对图片进行操作，在用dataloader之后才会对生成的图片进行数据增强，直接使用dataloader产生的数据就可以了。
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人:&nbsp;"Animesh Gupta"<notifications@github.com&gt;; 发送时间:&nbsp;2020年6月17日(星期三) 晚上10:51 收件人:&nbsp;"weiaicunzai/pytorch-cifar100"<pytorch-cifar100@noreply.github.com&gt;; 抄送:&nbsp;"夏洋洋"<1319230184@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;; 主题:&nbsp;Re: [weiaicunzai/pytorch-cifar100] 请问是不是models.resnet这些网络的实现有些问题？ (#17) 多谢哈，我这边也解决了，但是我发现是我的数据增强没有用上才导致精确度上不去的 您能告诉我如何解决这个问题，因为我无法获得相同的数字吗？ — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

@xyy1991 我没有对文件进行任何更改。 因此，现在我默认情况下使用的是“数据加载器”，您能建议我应该进行哪些更改以使我得到确切的错误吗？

**Comment by xyy1991** [comment]
这个如果没有实例是在不知道该怎么建议啊。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"Animesh Gupta"<notifications@github.com&gt;;
发送时间:&nbsp;2020年6月19日(星期五) 下午3:12
收件人:&nbsp;"weiaicunzai/pytorch-cifar100"<pytorch-cifar100@noreply.github.com&gt;;
抄送:&nbsp;"夏洋洋"<1319230184@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [weiaicunzai/pytorch-cifar100] 请问是不是models.resnet这些网络的实现有些问题？ (#17)





  
我自己实现的时候没有用dataloader，datasets虽然有transform这个属性，但是并没有对图片进行操作，在用dataloader之后才会对生成的图片进行数据增强，直接使用dataloader产生的数据就可以了。
 …
 ------------------&nbsp;原始邮件&nbsp;------------------ 发件人:&nbsp;"Animesh Gupta"<notifications@github.com&gt;; 发送时间:&nbsp;2020年6月17日(星期三) 晚上10:51 收件人:&nbsp;"weiaicunzai/pytorch-cifar100"<pytorch-cifar100@noreply.github.com&gt;; 抄送:&nbsp;"夏洋洋"<1319230184@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;; 主题:&nbsp;Re: [weiaicunzai/pytorch-cifar100] 请问是不是models.resnet这些网络的实现有些问题？ (#17) 多谢哈，我这边也解决了，但是我发现是我的数据增强没有用上才导致精确度上不去的 您能告诉我如何解决这个问题，因为我无法获得相同的数字吗？ — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.
  
@xyy1991 我没有对文件进行任何更改。 因此，现在我默认情况下使用的是“数据加载器”，您能建议我应该进行哪些更改以使我得到确切的错误吗？

**Comment by weiaicunzai** [comment]
> 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？

测试代码:
```
# my implementation
>>> from models.resnet import resnet50
>>> net = resnet50()
>>> sum(p.numel() for p in net.parameters())
23705252

# torchvision implementation
>>> from torchvision.models import resnet50
>>> net = resnet50()
>>> sum(p.numel() for p in net.parameters())
25557032
```
和官方大小差不多 , 你是怎么测的 5gb

**Comment by tjulyz** [comment]
我用的resnet发现64的时候gpu塞不下，占用显存过高，大家有这个问题吗？怎么解决的呢

**Comment by Patrickctyyx** [comment]
> > > > 我也有这个问题，作者的实现的差不多5个G,测试准确率有78.9， pytorch官方实现差不多1.4个G，测试集准确率也只有60%，，请问有看到问题在哪吗？
> > > > 而且pytorch官方实现速度快很多
> > > 
> > > 
> > > 我自己用官方的ResNet-18（改了下第一层卷积核大小适应cifar数据集尺寸）在cifar100的准确率是65%~70%之间。
> > > 我也尝试完全照搬作者的实验和整个工程去在cifar100上训练resnet18，但准确率只有66%，完全达不到>70%的准确率
> > > 不知道有没有和我一样情况的？无法复现作者的准确率？
> > > 我目前认为Cifar100只做数据增广，而不采用一些trick是无法让准确率上70%的
> > 
> > 
> > 请问您解决这个问题了吗？我好想遇到了跟你一样的情况
> 
> 我自己的问题解决了，我是因为pytorch版本原因导致了学习率没有正常更新，你检查下你的学习率是否在训练过程中更新，如果没有，尝试下把
> scheduler.step(epoch)
> 改成
> scheduler.step()
> 
> ps：resnet-18在cifar100上是可以达到76%左右的

感谢，不改的话第二个 ep 开始学习率就异常了

**Comment by Note-Liu** [comment]
Why the parameter 'epoch' in scheduler.step(epoch) will affect the lr ?
